### PR TITLE
UI and terminology guidelines in contributor's guide

### PR DESCRIPTION
**Pull request by oraNod**
https://issues.redhat.com/browse/ISPN-12384

### CODE REVIEW

**Comment by oraNod** [comment]
@karesti and @tristantarrant hey, I've been going through some style guidelines and wanted to add some details to our Contributor's Guide. there isn't a ton of stuff in here given that the PatternFly docs are pretty extensive and we don't need to duplicate all that. please let me know what you think. maybe we could fold this into the docs section.

**Review by karesti — APPROVED** [review]

**Comment by karesti** [comment]
I really like it! If Tristan is ok let's merge it

**Comment by oraNod** [comment]
@karesti thanks! I've realized we should add another section here. A simple table that maps terminology in code to recommended verbs. I'll work on it now and update. Adding the WIP label until it's all reviewed.

File: documentation/src/main/asciidoc/topics/contributing/ui.adoc [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+.Terminology
+
+* Use "Create" to mean build something new. For example, "create a cache".
+* Use "Add" to describe adding an existing item to an existing list, group, view, or other container element. For example, "add values to counters".
+
+|===
+|Method |Verb
+
+|`cache.put()`
+|Create
```
Comment (oraNod) [inline documentation/src/main/asciidoc/topics/contributing/ui.adoc:40]: @tristantarrant and @karesti As part of the Console UI review, we've got the recommendation to use "Create entries" instead of "Add entries". I'm not sure if I agree with that though. First you create the cache, then you add entries to the cache. I know it probably seems super trivial but it's important to make this consistent. Please let me know your thoughts. 

Also, I've gone through the REST API and some other docs. Are there any other methods that should have a corresponding verb in UI strings? I think it'd be helpful to have a list of what's in the code vs. how it should be presented via UI. Thanks.

**Review by danberindei — CHANGES_REQUESTED** [review]

File: documentation/src/main/asciidoc/topics/contributing/docs.adoc [hunk 2]
```diff
@@ -93,14 +93,24 @@ Underline (`[.underline]#value#`)::
 |for example  |e.g.
 |===
 
-Characters Per Line::
+.Terminology
+
+See the link:#terminology[{brandname} Terminology] section for guidance on using words and phrases specific to {brandname} capabilities and features.
+
+.Characters Per Line
```
[the remark this answers is not in the document]
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/contributing/docs.adoc:100]: yeah... let's just remove this one.

File: documentation/src/main/asciidoc/topics/contributing/docs.adoc [hunk 3]
```diff
@@ -93,14 +93,24 @@ Underline (`[.underline]#value#`)::
...
+.Asciidoc End of File
+
 To avoid formatting issues, end `adoc` files with two empty lines.
-Notes and Admonitions::
-Encapsulate notes, tips, and warnings in brackets. Surround the text in four equals characters, as follows:
-+
+
+. Notes and Admonitions
+
+Encapsulate notes, tips, and warnings in brackets. Surround the text in four
```
[the remark this answers is not in the document]
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/contributing/docs.adoc:111]: @danberindei sorry I don't follow your meaning. could you plz explain this one a little more?

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc [hunk 4]
```diff
@@ -0,0 +1,81 @@
+[[terminology]]
+= {brandname} Terminology
+
+{brandname} terminology refers to words and phrases specific to {brandname}
+capabilities and features. It is important to use consistent terminology across
+documentation and the code base to avoid unclear or contradictory language.
+
+Add::
+Add an item to an existing list, group, view, or other container element. For example, "add values to counters" or "add entries to caches".
```
[the remark this answers is not in the document]
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc:9]: right. it's more about incrementing counters. perhaps a moment of madness on my part. I'll remove that as an example.

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+A file in XML or JSON format that controls {brandname} behavior.
+
+Cache template::
+An artifact from which you create cache configurations. +
+Templates can be pre-defined cache configurations such as in the `org.infinispan.client.hotrod.DefaultTemplate` enum. +
+Users can use existing cache configurations as templates. +
+Users can also define templates in `infinispan.xml` with the `*-cache-configuration` tags.
+
+Cache definition::
+Do not use to mean "cache configuration".
```
[the remark this answers is not in the document]
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc:23]: This is one of the things that I intend on changing in the docs. I've noticed all the instances of "cache definition" too. I'm also guilty of using that phrase. It would be better to be consistent and say "cache configuration" instead. I think it would be fine to say "define a cache configuration" but I don't think we need two ways to say the one thing. 

I'll update the terminology docs to add definitions for these two.

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+Do not use to mean "delete".
+
+Index::
+Create an index to improve query performance.
+
+Indexes::
+Plural of index. Do not use "indices".
+
+Purge::
+Can have negative connotations and is not a preferred term. Use only as a synonym for "clear" in subsequent descriptions, such as in the body text of a dialog or text that follows a heading. Do not use as a primary verb or in headings.
```
[the remark this answers is not in the document]
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc:44]: For `purgeExpired()` would "delete" be better? I think it makes sense to expand "delete" to mean remove a single element or a group of related elements from a container.

**Review by ryanemerson — APPROVED** [review]

File: documentation/src/main/asciidoc/topics/contributing/docs.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (pruivo) [reply documentation/src/main/asciidoc/topics/contributing/docs.adoc:103]: @oraNod I think Dan is suggesting to each sentence have its own line, like:
```suggestion
Encapsulate notes, tips, and warnings in brackets.
Surround the text in four equals characters, as follows:
```
if a sentence is longer than 80 or 120 chars, it should be split into multiple lines.

[the remark this answers is not in the document]
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/contributing/docs.adoc:103]: @pruivo thanks! yeah I get it now. that's a good suggestion.

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+Cache configuration::
+An XML or JSON document that controls {brandname} behavior.
+
+Cache template::
+An artifact from which you create cache configurations. +
+{brandname} provides some pre-defined templates that can be referenced using the `org.infinispan.client.hotrod.DefaultTemplate` enum. +
+Users can define new templates in `infinispan.xml` with the `*-cache-configuration` tags. +
+In some APIs, you can use the configuration of an existing cache as a template to configure a new cache. +
+Users can also define templates in `infinispan.xml` with the `*-cache-configuration` tags.
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/contributing/terminology.adoc:19]: duplicated sentence (see line 17 above)

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+Add::
+Add an item to an existing list, group, view, or other container element. For example, "add entries to caches".
+
+Cache configuration::
+An XML or JSON document that controls {brandname} behavior.
+
+Cache template::
+An artifact from which you create cache configurations. +
+{brandname} provides some pre-defined templates that can be referenced using the `org.infinispan.client.hotrod.DefaultTemplate` enum. +
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/contributing/terminology.adoc:16]: `org.infinispan.client.hotrod.DefaultTemplate` is only for Hot Rod. I would mention that the REST and Server config could use the name instead. Example: `org.infinispan.DIST_SYNC`
It would be nice to have a link to DefaultTemplate.

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+Use in noun form preceded by Ickle: "Perform an Ickle query on the cache."
+
+Reindex::
+Rebuild an index to update or change an existing index. Do not hyphenate (always "reindex" never "re-index").
+
+Remove::
+Use as a synonym for "clear" or "delete" in subsequent descriptions, such as in the body text of a dialog or text that follows a heading. Do not use as a primary verb or in headings.
+
+Reset::
+Restore an initial value or state. For example, "Restore a counter."
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/contributing/terminology.adoc:62]: shouldn't the example be "reset a counter"?

File: documentation/src/main/asciidoc/topics/contributing/ui.adoc [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+This section describes guidelines for working with text in {brandname} user
+interface (UI) elements and applies, but is not limited, to the following
+components:
+
+* {brandname} Console
+* {brandname} CLI
+* log messages
+* error messages
+
+Content in the {brandname} UI should adhere to the link:https://www.patternfly.org/v4/design-guidelines/content/about[PatternFly content style guide].
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/contributing/ui.adoc:13]: link doesn't exist.

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+Add::
+Add an item to an existing list, group, view, or other container element. For example, "add entries to caches".
+
+Cache configuration::
+An XML or JSON document that controls {brandname} behavior.
+
+Cache template::
+An artifact from which you create cache configurations. +
+{brandname} provides some pre-defined templates that can be referenced using the `org.infinispan.client.hotrod.DefaultTemplate` enum. +
+Users can define new templates in `infinispan.xml` with the `*-cache-configuration` tags. +
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/contributing/terminology.adoc:17]: you forgot the `*` at the end of `-cache-configuration`
```suggestion
Users can define new templates in `infinispan.xml` with the `*-cache-configuration*` tags. +
```

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+Rebuild an index to update or change an existing index. Do not hyphenate (always "reindex" never "re-index").
+
+Remove::
+Use as a synonym for "clear" or "delete" in subsequent descriptions, such as in the body text of a dialog or text that follows a heading. Do not use as a primary verb or in headings.
+
+Reset::
+Restore an initial value or state. For example, "Restore a counter."
+
+Search::
+Synonymous with "query" but not preferred. In verb form, use with common phrases like "full-text search" in conjunction with {brandname} Query API. In noun form when referring to the "org.infinispan.query.Search" entry point.
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/contributing/terminology.adoc:65]: I would put `org.infinispan.query.Search` in monospace font instead of using quotes.

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
++
+Use in verb form: "Perform full-text searches through the {brandname} Query API."
+Use in noun form: "Use the {brandname} `Search` entry point to construct Ickle queries."
+
+== {brandname} methods and corresponding verbs
+
+When describing an action that the user performs that invokes an underlying method, use verbs as follows:
+
+|===
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/contributing/terminology.adoc:74]: is this table an example? or is it supposed to be a full list and include methods like `cache.remove()`?

**Review by pruivo — CHANGES_REQUESTED** [review]

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc (hunk 7 above)
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc:19]: fixed. thanks!

File: documentation/src/main/asciidoc/topics/contributing/ui.adoc (hunk 10 above)
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/contributing/ui.adoc:13]: weird. seems they changed it.

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc (hunk 13 above)
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc:74]: I've been debating if it would be worth the time to go through and do a full list. I started doing that but it seemed a little redundant with the preceding terminology list. I added `cache.remove()`. What do you think? Worth going through the full list?

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc (hunk 11 above)
Comment (pruivo) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc:17]: it looks like it is still missing :)

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc (hunk 13 above)
Comment (pruivo) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc:73]: nah, let's fill it lazily eh eh
when you find a wrong verb being used for a specific method, add it to the list :)

**Review by pruivo — APPROVED** [review]

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+{brandname} terminology refers to words and phrases specific to {brandname}
+capabilities and features. It is important to use consistent terminology across
+documentation and the code base to avoid unclear or contradictory language.
+
+Add::
+Add an item to an existing list, group, view, or other container element. For example, "add entries to caches".
+
+Cache configuration::
+An XML or JSON document that controls {brandname} behavior.
```
Comment (rigazilla) [inline documentation/src/main/asciidoc/topics/contributing/terminology.adoc:12]: Isn't this too general to define a cache configuration?

Comment (oraNod) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc:12]: @rigazilla Maybe but I'm not sure. What do you think is missing or how should we change the definition? 

For me, the "cache configuration" is the thing/artefact you use to enable features and define how you want Infinispan to behave. Which is distinct to a "configuration template" (a thing from which you create configurations). In docs I've seen configuration, template, and definition used inconsistently/interchangeably. Cheers.

Comment (danberindei) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc:12]: @oraNod I also think this definition is too broad. If we had a definition of "Cache" first, then the definition of "Cache configuration" could be "The specification of a cache".

OTOH it's not wide enough, because a configuration may be defined in a standalone XML/JSON document, as an element in the Infinispan embedded/server XML, or with Java code calling `ConfigurationBuilder` methods. Also we call a `Configuration` object a "cache configuration", even if it's not connected to an existing cache, or even registered in a cache manager.

In the end, I think the cache configuration/cache configuration template is a bit artificial right now, and we can't make this distinction as long as the same `DefaultCacheManager.defineConfiguration()` method is used to define both caches and templates.

Comment (danberindei) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc:12]: One more thing that probably needs to change for this to make sense: currently, a `Configuration` object does not include the cache name.

Perhaps that's why in some places we use "cache definition" instead of "cache configuration: Because "cache configuration" can be interpreted as the `Configuration` object, without the cache name, and XML document or element defining a cache does include the cache name, so it's somehow more than just a cache configuration.

/cc @tristantarrant @ryanemerson

Comment (oraNod) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc:12]: thanks for all those details @danberindei I need more time to digest that before making an update. IMO it's really good to start having this conversation to tackle some complexity. what you're saying helps point out the fact that what seems like a straightforward terminology definition isn't all that clear cut, which I think can come across to users.

Comment (gustavonalle) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc:12]: +1, XML or JSON is just about the format

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+Remove::
+Use as a synonym for "clear" or "delete" in subsequent descriptions, such as in the body text of a dialog or text that follows a heading. Do not use as a primary verb or in headings.
+
+Reset::
+Restore an initial value or state. For example, "Reset a counter."
+
+Search::
+Synonymous with "query" but not preferred. In verb form, use with common phrases like "full-text search" in conjunction with {brandname} Query API. In noun form when referring to the `org.infinispan.query.Search` entry point.
++
```
Comment (gustavonalle) [inline documentation/src/main/asciidoc/topics/contributing/terminology.adoc:65]: Better to not mention this by package, as we have 3 of them. Better yet, this is an implementation detail, I don't think it's notable to be part of this doc.

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc (hunk 14 above)
Comment (rigazilla) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc:12]: > @rigazilla Maybe but I'm not sure. What do you think is missing or how should we change the definition?
> 
> For me, the "cache configuration" is the thing/artefact you use to enable features and define how you want Infinispan to behave. Which is distinct to a "configuration template" (a thing from which you create configurations). In docs I've seen configuration, template, and definition used inconsistently/interchangeably. Cheers.

@oraNod forgive my delay, I've lost myself on another front.
I would add to the `cache definition` some more details to narrow down the part of Infinispan that a `cache definition` controls: i.e. we're going to control/represent a cache resource.
This is what it seems to me that is missing reading the definition.

Anyway @danberindei did a more detailed analysis above :)

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc (hunk 15 above)
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc:65]: Thanks. I've trimmed it down.

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc (hunk 14 above)
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc]: @danberindei and @rigazilla thanks, guys. I've updated this part. Honestly I'm of two minds about including terms for "cache configuration", "cache template", and "cache definition". Perhaps it doesn't make that much sense to create narrow definitions for these terms and we should remove them. 

The whole intent of this PR is to have general guidelines for improving consistency within UI (console, CLI, log messages) and documentation. Maybe I'm inventing an issue where it doesn't really exist but I feel like we should be able to tell users "when we say 'cache configuration' we mean x" instead of "we mean x but in some cases, we also mean y".

Comment (rigazilla) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc]: > @danberindei and @rigazilla thanks, guys. I've updated this part. Honestly I'm of two minds about including terms for "cache configuration", "cache template", and "cache definition". Perhaps it doesn't make that much sense to create narrow definitions for these terms and we should remove them.

@oraNod the new version looks better from my POV, IMO pragmatical guidelines are good enough for a contributing guide, if a sharp definition make things harder

File: documentation/src/main/asciidoc/topics/contributing/ui.adoc [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
++
+Heading in question form, starting with verb. ("Create new entry?") +
+Body that provides details and outcome of action. +
+Body that uses a synonym to the verb in heading where ambiguity might exist. +
+Buttons.
+* Use past tense form for success or failure messages.
++
+*Yes:* "Cache created." *No:* "Cache has been created."
+* Avoid using the word "please".
+* Ensure all icons and buttons with action have tooltip/hover help.
```
Comment (abigaeljamie) [inline documentation/src/main/asciidoc/topics/contributing/ui.adoc:28]: For accessibility reasons, we have on PatternFly to add a tooltip just for semantic icons (not decorative icons) that don't have labels (e.g., a trash can icon would have a tooltip reading "Delete"). A tooltip isn't needed for buttons or anything else with labels. So we could say something like, 

_For semantic icons without labels, add a label inside a tooltip that appears on hover. For more information on accessible icons, see [PatternFly's accessibility guide.](https://www.patternfly.org/v4/get-started/accessibility-guide)_

File: documentation/src/main/asciidoc/topics/contributing/ui.adoc [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+* Avoid abbreviations such as "avg" instead of "average".
```
Comment (abigaeljamie) [inline documentation/src/main/asciidoc/topics/contributing/ui.adoc:29]: We might also want to add "No." here since that appeared a lot. We should also avoid "e.g." and "i.e." Those sneak into UIs sometimes, but they're not easily understood by all users (especially those who speak a language without Latin roots). So maybe we could just expand this by saying something like, 

_Only use common and familiar abbreviations, such as "URL." Write out all other abbreviations, especially the following:_ 

_- Number: Avoid "No."_ 
_- Average: Avoid "avg."_ 
_- For example: Avoid "e.g."_ 
_- In other words: Avoid "i.e."_  

_For more information on abbreviations and acronyms, see [PatternFly's Numerics page in the UX writing style guide.](https://www.patternfly.org/v4/ux-writing/terminology)_

File: documentation/src/main/asciidoc/topics/contributing/ui.adoc [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+This section describes guidelines for working with text in {brandname} user
+interface (UI) elements and applies, but is not limited, to the following
+components:
+
+* {brandname} Console
+* {brandname} CLI
+* log messages
+* error messages
+
+Content in the {brandname} UI should adhere to the link:https://www.patternfly.org/v4/ux-writing/about[PatternFly content style guide].
```
Comment (abigaeljamie) [inline documentation/src/main/asciidoc/topics/contributing/ui.adoc:13]: We call this the "PatternFly UX writing style guide." The name matches the nav item so that it's easier to find 👍

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+Look up information within the data set. The words "query" and "search" are synonymous but frequently interchanging the two can lead to inconsistency.
++
+Use in verb form: "Query values in the cache." +
+Use in noun form preceded by Ickle: "Perform an Ickle query on the cache."
+
+Reindex::
+Rebuild an index to update or change an existing index. Do not hyphenate (always "reindex" never "re-index").
+
+Remove::
+Use as a synonym for "clear" or "delete" in subsequent descriptions, such as in the body text of a dialog or text that follows a heading. Do not use as a primary verb or in headings.
```
Comment (abigaeljamie) [inline documentation/src/main/asciidoc/topics/contributing/terminology.adoc:64]: This contradicts the PatternFly terminology list since "delete" and "remove" mean different things: 

Remove: (v.) Use to describe removing an item from a list, group, view, or other container element without completely deleting it. Also see Add and Delete.

Opposite: Add
Notes: If what you’re removing is not readily apparent from the context, consider adding a noun (like “Remove file”). 

We may not want to encourage synonyms since the inconsistency has potential to confuse users (more info in a later comment).

**Review by abigaeljamie — COMMENTED** [review]
Great job on these guidelines, @oraNod ! I left some comments throughout for you to accept/reject as you see fit 👍 Thank you for including me in this process and for creating a helpful resource for future projects!

File: documentation/src/main/asciidoc/topics/contributing/terminology.adoc (hunk 19 above)
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc:64]: Thanks Abi. FYI, I added this based on some notes from our doc so we might want to revisit that before we submit any PRs for the console.

"Remove: a synonym for “clear” or “delete” in the body copy/dialog"

There are also some places, particularly in the REST API, where "remove" is used synonymously with "delete". TBH I'm not sure if the distinction between "remove" and "delete" that you're making will apply to Infinispan.

Generally, when we say "remove" something it's deleted, rather than take it out of a container but it still exists. We even use "remove" in the definition for "delete" in this terminology list.

What if we advise using "permanently remove" in places where there is a need to describe "delete"?

Comment (abigaeljamie) [reply documentation/src/main/asciidoc/topics/contributing/terminology.adoc]: Ah, got it. It was one of the verbs you, Alana, and Ingrid defined (that was the day I had to drop the call early). If you feel like it makes sense as a synonym and it suits Infinispan best, then let's go with it 👍

**Review by rigazilla — APPROVED** [review]
